include/ginkgo/core/matrix/csr.hpp

```diff
@@ -231,34 +343,84 @@ class Csr : public EnableLinOp<Csr<ValueType, IndexType>>,
 
         int64_t clac_size(const int64_t nnz)
         {
-            constexpr uint32 warp_size = 32;
-            int multiple = 8;
-            if (nnz >= 2000000) {
-                multiple = 128;
-            } else if (nnz >= 200000) {
-                multiple = 32;
+            if (warp_size_ > 0) {
+                int multiple = 8;
+                if (nnz >= 2e6) {
+                    multiple = 128;
+                } else if (nnz >= 2e5) {
+                    multiple = 32;
+                }
+
+#if GINKGO_HIP_PLATFORM_HCC
+                if (!cuda_strategy_) {
+                    multiple = 8;
+                    if (nnz >= 1e7) {
+                        multiple = 64;
+                    } else if (nnz >= 1e6) {
+                        multiple = 16;
+                    }
+                }
+#endif  // GINKGO_HIP_PLATFORM_HCC
+
+                auto nwarps = nwarps_ * multiple;
+                return min(ceildiv(nnz, warp_size_), int64_t(nwarps));
+            } else {
+                return 0;
             }
-            auto nwarps = nwarps_ * multiple;
-            return min(ceildiv(nnz, warp_size), static_cast<int64_t>(nwarps));
         }
 
     private:
         int64_t nwarps_;
+        int warp_size_;
+        bool cuda_strategy_;
     };
 
     class automatical : public strategy_type {
     public:
+        /**
+         * Creates a automatical strategy.
+         */
         automatical()
             : automatical(std::move(
                   gko::CudaExecutor::create(0, gko::OmpExecutor::create())))
         {}
 
+        /**
+         * Creates a automatical strategy with CUDA executor.
+         *
+         * @param exec the CUDA executor
+         */
         automatical(std::shared_ptr<const CudaExecutor> exec)
-            : automatical(exec->get_num_warps())
+            : automatical(exec->get_num_warps(), exec->get_warp_size())
+        {}
+
+        /**
+         * Creates a automatical strategy with HIP executor.
+         *
+         * @param exec the HIP executor
+         */
+        automatical(std::shared_ptr<const HipExecutor> exec)
+            : automatical(exec->get_num_warps(), exec->get_warp_size(), false)
         {}
 
-        automatical(int64_t nwarps)
-            : strategy_type("automatical"), nwarps_(nwarps)
+        /**
+         * Creates a automatical strategy with specified parameters
+         *
+         * @param nwarps the number of warps in the executor
+         * @param warp_size the warp size of the executor
+         * @param cuda_strategy whether uses the cuda_strategy.
+         *
+         * @note The apply results are not unspecified when the warp_size is not
+         *       correct. This information may be lost in conversion. To avoid
+         *       it, needs to use `set_strategy` to set the strategy with
+         *       correct parameters.
```

Comment:
This triple negative is a bit confusing. Do you mean to say that when a matrix is converted, the strategy is lost and hence the warp_size is not correctly set ? Then I would re-frame the sentence like this

 " When converting to CSR, `set_strategy` needs to be called with correct parameters to properly set the `warp_size` parameter which is unset during the conversion. " or something like this if i misunderstood what you meant to say.